File: src/test/java/io/tolgee/controllers/SignedInControllerTest.java
package io.tolgee.controllers;

import io.tolgee.helpers.JsonHelper;
import io.tolgee.model.UserAccount;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.ResultActions;
import org.testng.annotations.BeforeMethod;

import static io.tolgee.controllers.LoggedRequestFactory.*;

@SpringBootTest
@AutoConfigureMockMvc
public abstract class SignedInControllerTest extends AbstractControllerTest {
    UserAccount userAccount;

    @BeforeMethod
    public void beforeEach() throws Exception {
        //populate to create the user if not created
        dbPopulator.autoPopulate();
        if (userAccount == null) {
            logAsUser(tolgeeProperties.getAuthentication().getInitialUsername(), initialPassword);
        }
        commitTransaction();
    }

    public void logAsUser(String userName, String password) throws Exception {
        DefaultAuthenticationResult defaultAuthenticationResult = login(userName, password);
        LoggedRequestFactory.init(defaultAuthenticationResult.getToken());
        this.userAccount = defaultAuthenticationResult.getEntity();
    }

    public void logout() {
        userAccount = null;
    }

    public ResultActions performPost(String url, Object content) {
        try {
            return mvc.perform(loggedPost(url).contentType(MediaType.APPLICATION_JSON).content(JsonHelper.asJsonString(content)));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public ResultActions performGet(String url) {
        try {
            return mvc.perform(loggedGet(url));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public ResultActions performDelete(String url, Object content) {
        try {
            return mvc.perform(loggedDelete(url).contentType(MediaType.APPLICATION_JSON).content(JsonHelper.asJsonString(content)));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

}

File: src/test/kotlin/io/tolgee/controllers/SignedInControllerTest.kt
package io.tolgee.controllers

import io.tolgee.fixtures.AuthRequestPerformer
import io.tolgee.fixtures.LoggedRequestFactory.init
import io.tolgee.fixtures.SignedInRequestPerformer
import io.tolgee.model.UserAccount
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.web.servlet.ResultActions
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder
import org.testng.annotations.AfterMethod
import org.testng.annotations.BeforeMethod

@SpringBootTest
@AutoConfigureMockMvc
abstract class SignedInControllerTest : AbstractControllerTest(), AuthRequestPerformer {
    var userAccount: UserAccount? = null

    @Autowired
    lateinit var signedInRequestPerformer: SignedInRequestPerformer

    @BeforeMethod
    fun beforeEach() {
        //populate to create the user if not created
        dbPopulator.createUserIfNotExists(tolgeeProperties.authentication.initialUsername)
        if (userAccount == null) {
            logAsUser(tolgeeProperties.authentication.initialUsername, initialPassword)
        }
        commitTransaction()
    }

    @AfterMethod
    fun afterEach() {
        logout()
    }

    fun logAsUser(userName: String, password: String) {
        val defaultAuthenticationResult = login(userName, password)
        init(defaultAuthenticationResult.token)
        userAccount = defaultAuthenticationResult.entity
    }

    fun logout() {
        userAccount = null
    }

    override fun perform(builder: MockHttpServletRequestBuilder): ResultActions {
        return requestPerformer.perform(builder)
    }

    override fun performDelete(url: String, content: Any?): ResultActions {
        return requestPerformer.performDelete(url, content)
    }

    override fun performGet(url: String): ResultActions {
        return requestPerformer.performGet(url)
    }

    override fun performPost(url: String, content: Any?): ResultActions {
        return requestPerformer.performPost(url, content)
    }

    override fun performPut(url: String, content: Any?): ResultActions {
        return requestPerformer.performPut(url, content)
    }

    override fun performAuthPut(url: String, content: Any?): ResultActions {
        return signedInRequestPerformer.performAuthPut(url, content)
    }

    override fun performAuthPost(url: String, content: Any?): ResultActions {
        return signedInRequestPerformer.performAuthPost(url, content)
    }

    override fun performAuthGet(url: String): ResultActions {
        return signedInRequestPerformer.performAuthGet(url)
    }

    override fun performAuthDelete(url: String, content: Any?): ResultActions {
        return signedInRequestPerformer.performAuthDelete(url, content)
    }
}
